crates/subspace-networking/src/behavior/persistent_parameters.rs

```diff
@@ -0,0 +1,248 @@
+            *addresses = addresses.union(&addr_set).cloned().collect()
+        } else {
+            self.known_peers.push(peer_id, addr_set);
+        }
+    }
+
+    async fn known_addresses(&self, peer_number: usize) -> Vec<(PeerId, Multiaddr)> {
+        self.known_peers
+            .iter()
+            .take(peer_number)
+            .flat_map(|(peer_id, addresses)| addresses.iter().map(|addr| (*peer_id, addr.clone())))
+            .map(|(peer_id, addr)| {
+                // remove p2p-protocol suffix if any
+                let mut modified_address = addr.clone();
+
+                if let Some(Protocol::P2p(_)) = modified_address.pop() {
+                    (peer_id, modified_address)
+                } else {
+                    (peer_id, addr)
+                }
+            })
+            .collect()
+    }
+
+    async fn run(&mut self) {
+        (&mut self.networking_parameters_save_delay).await;
+
+        // save accumulated cache to DB
+        let dto: NetworkingParametersDto = self.clone_known_peers().into();
+        let save_result = serde_json::to_vec(&dto)
+            .map_err(NetworkParametersPersistenceError::from)
+            .and_then(|data| {
+                let tx = vec![(self.column_id, self.object_id, Some(data))];
+
+                self.db.commit(tx).map_err(|err| err.into())
+            });
+
+        if let Err(err) = save_result {
+            debug!(error=%err, "Error on saving network parameters");
+        } else {
+            trace!("Networking parameters saved to DB");
+        }
+
+        self.networking_parameters_save_delay = NetworkingParametersManager::default_delay();
+    }
+
+    fn clone_box(&self) -> Box<dyn NetworkingParametersRegistry> {
+        Self {
+            known_peers: self.clone_known_peers(),
+            networking_parameters_save_delay: Self::default_delay(),
+            db: self.db.clone(),
+            column_id: self.column_id,
+            object_id: self.object_id,
+        }
+        .boxed()
+    }
+}
+
+impl From<&LruCache<PeerId, HashSet<Multiaddr>>> for NetworkingParametersDto {
+    fn from(cache: &LruCache<PeerId, HashSet<Multiaddr>>) -> Self {
+        Self {
+            known_peers: cache
+                .iter()
+                .map(|(peer_id, addresses)| (*peer_id, addresses.clone()))
+                .collect(),
+        }
+    }
+}
+
+impl From<LruCache<PeerId, HashSet<Multiaddr>>> for NetworkingParametersDto {
+    fn from(cache: LruCache<PeerId, HashSet<Multiaddr>>) -> Self {
+        From::<&LruCache<PeerId, HashSet<Multiaddr>>>::from(&cache)
+    }
+}
+
+impl From<NetworkingParametersDto> for LruCache<PeerId, HashSet<Multiaddr>> {
+    fn from(params: NetworkingParametersDto) -> Self {
+        let mut known_peers = LruCache::<PeerId, HashSet<Multiaddr>>::new(PEER_CACHE_SIZE);
+
+        for (peer_id, addresses) in params.known_peers.iter() {
+            known_peers.push(*peer_id, addresses.clone());
+        }
+
+        known_peers
+    }
+}
```

Comment:
Why are these conversions necessary? Feels like a strange and too specific kind of conversion.

Reply:
These seem useful for reusable cases, with private data structure (BTW, should be possible to rename to just `NetworkingParameters`, we don't really use `Dto` suffix anywhere regardless of what struct represents) should be fine to just create an instance explicitly.